x-vector/src/X/Data/Vector/Stream/Group.hs

```diff
@@ -32,39 +33,60 @@ concatWith =
   flip Stream.concatMap
 {-# INLINE [1] concatWith #-}
 
-chunksOf :: (PrimMonad b, MonadBase b m, Generic.Vector v a) => Int -> Stream m a -> Stream m (v a)
-chunksOf !n (Stream step t) =
+-- | Turns a stream of @a@ in to a stream of chunks of size @n@.
+chunked :: (PrimMonad b, MonadBase b m, Generic.Vector v a) => Int -> Stream m a -> Stream m (v a)
+chunked n xs =
   if n <= 0 then
-    Savage.error "X.Data.Vector.Stream.Group.chunksOf: chunk size must be greater than zero"
+    Savage.error "X.Data.Vector.Stream.Group.chunked: chunk size must be greater than zero"
+  else
+    grouped (\_ _ -> False) n xs
+{-# INLINE [1] chunked #-}
+
+-- | Turns a stream of @a@ in to a stream of chunks of at least size @n@,
+--   except for the last one. Values of @a@ which are equal according to the
+--   comparison function stay in the same chunk.
+grouped :: (PrimMonad b, MonadBase b m, Generic.Vector v a) => (a -> a -> Bool) -> Int -> Stream m a -> Stream m (v a)
+grouped eq n (Stream step t) =
+  if n <= 0 then
+    Savage.error "X.Data.Vector.Stream.Group.grouped: chunk size must be greater than zero"
   else
     embed $ do
       let
+        notEq mx y =
+          case mx of
+            Nothing' ->
+              True
+            Just' x ->
+              not (eq x y)
+        {-# INLINE [0] notEq #-}
+
         loop = \case
           Nothing' ->
             pure Done
 
-          Just' (s0, i, g0) ->
-            if i == n then do
-              xs <- liftBase $ Grow.unsafeFreeze g0
-              g <- liftBase $ Grow.new n
-              pure . Yield xs $ Just' (s0, 0, g)
-            else
-              step s0 >>= \case
-                Yield x s -> do
+          Just' (s0, i, last, g0) ->
+            step s0 >>= \case
+              Yield x s ->
+                if i >= n && last `notEq` x then do
+                  xs <- liftBase $ Grow.unsafeFreeze g0
+                  g <- liftBase $ Grow.new n
+                  liftBase $ Grow.add g x
+                  pure . Yield xs $ Just' (s, 1, Just' x, g)
+                else do
                   liftBase $ Grow.add g0 x
-                  pure . Skip $ Just' (s, i + 1, g0)
-                Skip s ->
-                  pure . Skip $ Just' (s, i, g0)
-                Done ->
-                  if i == 0 then
-                    pure $ Skip Nothing'
-                  else do
-                    xs <- liftBase $ Grow.unsafeFreeze g0
-                    pure $ Yield xs Nothing'
+                  pure . Skip $ Just' (s, i + 1, Just' x, g0)
```

Comment:
Yeah, that's right.

I was going to go the other way and try to keep the chunk size under `n`, but unfortunately if you have more than `n` values of `a` you're stuck. So I thought it was simpler to have chunks of at least `n` rather than at most `n`.